File: design-patterns/java/DesignPatterns/src/ar/edu/unlam/cdp/patterns/command/CompositeCommand.java
package ar.edu.unlam.cdp.patterns.command;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class CompositeCommand extends Command {

	List<Command> commands = new LinkedList<Command>();
	
	CompositeCommand() {
		super("");
	}

	public void add(Command command) {
		this.commands.add(command);
	}
	
	@Override
	public void execute() {
		for (Command command : this.commands) {
			command.execute();
		}
	}

	@Override
	public void undo() {
		Collections.reverse(this.commands);
		for (Command command : this.commands) {
			command.execute();
		}
		Collections.reverse(this.commands);
	}

	@Override
	public String getDescription(){
		String desc = "";
		for (Command command : this.commands) {
			desc = desc + command.getDescription() + "\n";
		}
		return desc;
	}
	
}

File: design-patterns/cs/Patrones/Patrones/Command/CompositeCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Patrones.Command
{
    public class CompositeCommand : Command
    {
        IList<Command> commands = new List<Command>();

        public CompositeCommand() : base("") { }

        public void Add(Command command)
        {
            commands.Add(command);
        }

        #region Metodos Abstractos Implementados

        public override void Execute()
        {
            foreach (var command in commands)
            {
                command.Execute();
            }
        }

        public override void Undo()
        {
            foreach (var command in commands.Reverse())
            {
                command.Undo();
            }
        }

        public new string Description
        {
            get
            {
                string aux = string.Empty;
                foreach (var command in commands)
                {
                    aux += command.Description + Environment.NewLine;
                }

                return aux;
            }
        }

        #endregion
    }
}
